src/server/pfs/server/driver_v2.go

```diff
@@ -334,13 +370,18 @@ func (d *driverV2) compactIter(ctx context.Context, params compactSpec) (_ *comp
 	// TODO: use an errgroup to make the recursion concurrecnt.
 	// this requires changing the master to allow multiple calls to RunSubtasks
 	// don't forget to pass the errgroups childCtx to compactIter instead of ctx.
+	// TODO: change this such that the fan in is maxed at the lower levels first rather
+	// than the higher.
 	for i := 0; i < params.maxFanIn; i++ {
 		start := i * childSize
+		if start >= len(params.inputPaths) {
+			break
+		}
 		end := (i + 1) * childSize
 		if end > len(params.inputPaths) {
 			end = len(params.inputPaths)
 		}
-		childOutputPath := path.Join(scratch, strconv.Itoa(i))
+		childOutputPath := path.Join(scratch, fileset.SubFileSetStr(int64(i)))
```

Comment:
This wasn't a bug before, the list ensures ordering, we don't have to rely on lexicographical ordering.

Reply:
The issue is the common prefix, not the ordering. So, when setting up the merge reader we would end up grabbing some filesets more than once (i.e. greater than 10 paths would result with duplicate data because the 10th fileset would be grabbed twice, once by the walk of 1 and once by the walk of 10). This was causing data to be duplicated.